#190 GlobalsToArguments transformation

sergisiso opened the pull request:
Add a transformation to promote global variables like `use my_mod, only: parameter1` to an argument with the global access in the caller (needed in OpenCL and OpenACC).

It always needs the global variable type for the subroutine signature, so if the type is deferred (common in global vars) it has to be evaluated with the functionality introduced by #545

sergisiso commented:
@rupertford @arporter This is ready for the first review, it implements the transformation to convert kernel globals to arguments. It is implemented in a generic way (for all APIs) but I tested it mainly on the nemolite2d (Gocean) code. For the time being this relies on InvokeSchedule having a SymbolTable and a NameSpaceManager coexisting.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -4368,3 +4368,88 @@ def validate(self, node_list, options=None):
+            current_arg_list.append(globalvar)
+            symtab.specify_argument_list(current_arg_list)
+
+            # Add the Global variable in the call argument list
```
This section feels like it belongs in a new method e.g. `node.arguments.add(type, name, access)`, i.e. create a new method in the psyGen.Arguments base class. Is that do-able?

sergisiso replied:
Done, but I used `append` as method name as I just used it to add at the end of the argument list.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1809,9 +1809,20 @@ def __init__(self, KernFactory, BuiltInFactory, alg_calls=None):
         # InvokeSchedule opencl_options default values
         self._opencl_options = {"end_barrier": True}
+        # TODO: #312 Currently NameSpaceManager and SymbolTable coexist, but
+        # it would be better to merge them
```
Please also mention that the SymbolTable is currently only used to manage the import of global symbols that were previously inside kernels.

sergisiso replied:
Extended comment as suggested

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1878,6 +1889,10 @@ def gen_code(self, parent):
             IfThenGen, CallGen
 
+        for globalvar in self.symbol_table.global_datasymbols:
+            parent.add(UseGen(parent,
```
Please also add a note here to say that currently we only use the SymbolTable for global symbols promoted from kernels.

sergisiso replied:
Done both. Produce a single statement per container_symbol and added comment as suggested.

sergisiso commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -358,11 +358,11 @@ def constant_value(self, new_value):
-                    "Error setting '{0}' constant value. A DataSymbol with a "
-                    "constant value is currently limited to Local interfaces "
-                    "but found '{1}'.".format(self.name, self.interface))
+                    "Error setting '{0}' constant value. A DataSymbol with "
```
Agreed, I updated this one and the other similar error messages.

arporter replied:
Thanks.

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -337,6 +338,54 @@ def data_arguments(self):
+        :param globalvar: the variable to be copied in
+        :type globalvar: :py:class:`psyclone.psyir.symbols.DataSymbol`
+
+        :raise TypeError: if the given variable is not a global variable
```
Ups, fixed

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -337,6 +338,54 @@ def data_arguments(self):
+        :raise KeyError: if the given variable name already exists in the \
+            symbol table
+        '''
+
```
Deleted

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -337,6 +338,54 @@ def data_arguments(self):
+        if external_container_name not in self:
+            self.add(ContainerSymbol(external_container_name))
+
+        # Copy the variable in the SymbolTable with the appropriate interface
```
Done

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -337,6 +338,54 @@ def data_arguments(self):
+            if not (self.lookup(globalvar.name).is_global and
+                    self.lookup(globalvar.name).interface
+                    .container_symbol.name == external_container_name):
+                raise KeyError(
```
I thought about it being a SymbolError but the `add` method also returns a KeyError when the name already exists in the SymbolTable and I think it should be the same error.

arporter replied:
OK.

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -337,6 +338,54 @@ def data_arguments(self):
+            raise TypeError(
+                "The globalvar argument of SymbolTable.copy_external_"
+                "global method should have a GlobalInterface interface, "
+                "but found {0}.".format(type(globalvar.interface)))
```
Since the rest of the code mostly uses `type(xxx).__name__` I will use this in all the introduced error messages, but I am not sure if I agree that it's nicer :)

sergisiso commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -428,3 +428,63 @@ def test_symboltable_unresolved():
+        symtab.copy_external_global(DataSymbol("var1", "real"))
+    assert "The globalvar argument of SymbolTable.copy_external_global " \
+        "method should have a GlobalInterface interface, but found <class " \
+        "'psyclone.psyir.symbols.datasymbol.LocalInterface'>." \
```
Done

arporter commented:
I had a further thought after I'd finished reviewing - what happens if the global accessed by a kernel turns out to be a parameter? In that case I suspect most compilers will object to us passing it as an argument with 'INOUT' intent.

sergisiso commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -428,3 +428,63 @@ def test_symboltable_unresolved():
+        symtab.lookup("b").interface.container_symbol
+
+    # The copy of globalvars that already exist is supported
+    var3 = DataSymbol("c", "deferred", interface=GlobalInterface(container2))
```
Well spotted, this (and the one below) should have been "b"  (note that the ContainerSymbol is already present is checked in the set of tests before this one)

sergisiso commented on src/psyclone/transformations.py:
```diff
@@ -4368,3 +4368,88 @@ def validate(self, node_list, options=None):
+
+
+class KernelGlobalsToArguments(Transformation):
+    '''Transformation that given a kernel call, it takes out of the
```
Done

sergisiso commented on src/psyclone/transformations.py:
```diff
@@ -4368,3 +4368,88 @@ def validate(self, node_list, options=None):
+
+    @property
+    def name(self):
+        ''' Returns the name of this transformation as a string.'''
```
Done

sergisiso commented on src/psyclone/transformations.py:
```diff
@@ -4368,3 +4368,88 @@ def validate(self, node_list, options=None):
+        '''
+        Check that the supplied node is a valid target for this transformation.
+
+        :param node: the loop node to validate.
```
I made it: "the PSyIR node to validate."

sergisiso commented on src/psyclone/transformations.py:
```diff
@@ -4368,3 +4368,88 @@ def validate(self, node_list, options=None):
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :raises TransformationError: if the supplied node is not a CodedKern
```
Done

sergisiso commented on src/psyclone/transformations.py:
```diff
@@ -4368,3 +4368,88 @@ def validate(self, node_list, options=None):
+            raise TransformationError(
+                "The KernelGlobalsToArguments transformation can "
+                "only be applied to CodedKern nodes but found {0} instead."
+                "".format(type(node)))
```
Done

sergisiso commented on src/psyclone/transformations.py:
```diff
@@ -4368,3 +4368,88 @@ def validate(self, node_list, options=None):
+        modify the InvokeSchedule to pass the same global variables to the
+        kernel call.
+
+        :param node: a kernel call
```
Fixed

sergisiso commented:
@arporter This PR is ready for the next review. I addressed the in-line comments but also end up refactoring part of the code. A relevant modification is that I constrained the transformation to only the GOcean API (at least for now) and therefore it adds GOKernelArguments instead of generic Arguments.

I also added more tests scenarios with multiple kernels and possible variable clashes. However, due to the limitations in the test compilation infrastructure I still don't try to compile the outputs (I made a new issue about this).

arporter commented on src/psyclone/tests/test_files/gocean1p0/model_mod.f90:
```diff
@@ -30,12 +29,14 @@
 ! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
-! Author J.Henrichs, Bureau of Meteorology
+! Author: J.Henrichs, Bureau of Meteorology
```
Please update the copyright date on this file.

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -4097,27 +3810,34 @@ def validate(self, node, options=None):
         :type options: dictionary of string:values or None
 
-        :raises TransformationError: if the supplied node is not a CodedKern
+        :raises TransformationError: if the supplied node is not a CodedKern.
```
Please also document the exception raised if we're not using the NEMO API.

sergisiso replied:
Done. (with Gocean API you mean)

arporter replied:
Oops, yes and thanks.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -4097,27 +3810,34 @@ def validate(self, node, options=None):
+
+        if not isinstance(node.root, GOInvokeSchedule):
+            raise TransformationError(
+                "The {0} generation is currently only supported for the "
```
s/generation/transformation/ I think?

sergisiso replied:
Yep

arporter commented on src/psyclone/gocean1p0.py:
```diff
@@ -1468,6 +1468,26 @@ def scalars(self):
         return [arg.name for arg in args]
 
+    def append(self, name):
+        ''' Append a GOKernelArgument to the Argument list.
```
"Create and append a GOKernelArgument..."

sergisiso replied:
Done

arporter commented on src/psyclone/tests/gocean1p0_test.py:
```diff
@@ -1476,3 +1478,280 @@ def test14_no_builtins():
+    assert argument.type == "descriptor_type"
+
+
+def test_globalstoargumentstrans_wrongapi():
```
This and the following transformation tests should be somewhere else. Probably `tests/domain/gocean/transformations/globalstoargs_test.py` or some such.

sergisiso replied:
Moved tests to the proposed file

sergisiso commented:
@arporter Ready for next review

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+    notkernel = invoke.schedule.children[0]
+    kernel = invoke.schedule.coded_kernels()[0]
+
+    # Monckeypatch resolve_deferred to avoid module searching and importing
```
"Monkeypatch" and I think you can monkeypatch the class itself rather than having to separately do it for each DataSymbol object. You might need to do it before those objects are created though.

sergisiso replied:
Ah true, monkeypatching the class worked well with a simpler code and solved the pylint issue.

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+
+
+def test_globalstoargumentstrans(monkeypatch):
+    ''' Check the KernelGlobalsToArguments transformation '''
```
Please make this docstring a bit more specific.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+
+
+def test_globalstoargumentstrans_constant(monkeypatch):
+    ''' Check the KernelGlobalsToArguments transformation '''
```
Please make this docstring more explicit/give more detail.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+    from psyclone.psyir.backend.fortran import FortranWriter
+
+    trans = KernelGlobalsToArguments()
+    assert trans.name == "KernelGlobalsToArguments"
```
You already have these checks in the previous test :-)

sergisiso replied:
Removed from here

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+    invoke = psy.invokes.invoke_list[0]
+    kernel = invoke.schedule.coded_kernels()[0]
+
+    # Monckeypatch resolve_deferred to avoid module searching and importing
```
"Monkeypatch" and again, I think you can patch the class rather than all of the separate instances.

sergisiso replied:
Fixed the same way as in the previous test.

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+    # Monckeypatch resolve_deferred to avoid module searching and importing
+    # in this test. In this case we assume it is a constant INTEGER
+    for var in kernel.get_kernel_schedule().symbol_table.global_datasymbols:
+        def set_to_real(variable):
```
Should be `set_to_integer`.

sergisiso replied:
Ups, renamed to `set_to_const_int`

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+    for num, kernel in enumerate(invoke.schedule.coded_kernels()):
+        kschedule = kernel.get_kernel_schedule()
+
+        # We already tested that DEFERRED globals are resolved in last test,
```
This comment isn't quite true any more. It's tested somewhere but not in the last test.

sergisiso replied:
Remove this part of the sentence

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+    kernel = invoke.schedule.coded_kernels()[0]
+    trans = KernelGlobalsToArguments()
+    before_code = str(psy.gen)
+    trans.apply(kernel)
```
Please could you add a TODO here referencing #11 as we may want to warn the user that no globals were found. In fact, could you add it in the `apply()` method too?

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -3785,3 +3785,101 @@ def validate(self, loop, options=None):
+        invoke_symtab = node.root.symbol_table
+
+        # Transform each global variable into an argument
+        for globalvar in kernel.symbol_table.global_datasymbols:
```
Please add a TODO here for #11 to generate a warning/info message if there aren't any global datasymbols in this kernel.

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -3785,3 +3785,101 @@ def validate(self, loop, options=None):
+            # Convert the symbol to an argument and add it to the argument list
+            current_arg_list = symtab.argument_list
+            if globalvar.is_constant:
+                # Global constants lose the constant value but are read-only
```
Please add a TODO here to say that a further transformation will be introduced to support replacing global constants with local ones (and create an issue if we don't already have one).

sergisiso replied:
Done

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+# -----------------------------------------------------------------------------
```
I think it's my mistake but please could you put this file in `tests/domain/gocean/transformations`?

sergisiso replied:
Done

arporter commented on src/psyclone/tests/gocean/transformations/globalstoargs_test.py:
```diff
@@ -0,0 +1,260 @@
+
+    # Monckeypatch resolve_deferred to avoid module searching and importing
+    # in this test. In this case we assume it is a REAL
+    for var in kernel.get_kernel_schedule().symbol_table.global_datasymbols:
```
This loop and the similar one at line 170 generate pylint warnings about "cell variable var defined in loop". Please could you check what this means and see whether it can be avoided (in fact, my suggestion above about monkeypatching the class might remove the need for this loop).

sergisiso replied:
Fixed by your suggestion above.

arporter commented on src/psyclone/tests/gocean1p0_test.py:
```diff
@@ -45,7 +45,8 @@
 from psyclone.generator import GenerationError, ParseError
-from psyclone.gocean1p0 import GOKern, GOLoop, GOInvokeSchedule
+from psyclone.gocean1p0 import GOKern, GOLoop, GOInvokeSchedule, \
+    GOKernelArgument, GOKernelArguments
```
pylint complains that `GOKernelArgument` is subsequently re-imported further down.

sergisiso replied:
Deleted the duplicated import inside the test

sergisiso commented:
@arporter This is ready for the next review. Inline comments have been addressed.

Regarding the name clashes there is 2 possibilities:
- Clashes with already defined symbols in the symbol table. This produce an error in the transformation.
- Clashes with other NameSpace variables. This do not happen during the transformation as the NameSpace is populated during the generation phase. There is 2 cases, if the name was already used it breaks because the global var should be an exact match. If the name is used latter, it just follows the namespace indexing convention without producing an Error.
I added new tests for all of these cases.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1924,11 +1924,30 @@ def gen_code(self, parent):
+                    context="AlgArgs",
+                    label=var_name)
+                # There is a name clash with this variable name and we can not
+                # accept indexed names for global variables.
```
Please add a TODO (and new issue if necessary) here as we should try to work around this. e.g. in Fortran you can re-name the variable being imported with a USE. However, that wouldn't help us if the module name itself clashed.

sergisiso replied:
Added TODO and created issue #642

arporter commented on src/psyclone/tests/domain/gocean/transformations/globalstoargs_test.py:
```diff
@@ -45,7 +45,7 @@
 
 BASEPATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
-                        "..", "..", "test_files")
+                        "..", "..", "..", "test_files")
```
Slightly picky but we shouldn't do ".." to go up a directory. Please use functionality from within `os.path` to do this (there are examples of this at a few places in the code).

sergisiso replied:
Done

arporter commented on src/psyclone/tests/domain/gocean/transformations/globalstoargs_test.py:
```diff
@@ -91,10 +93,9 @@ def test_globalstoargumentstrans(monkeypatch):
 
     # Monckeypatch resolve_deferred to avoid module searching and importing
```
It's a small thing but you've consistently spelt monkeypatch as "monckeypatch" in this file :-) Please could you fix.

sergisiso replied:
Ups.. fixed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -3872,6 +3874,8 @@ def apply(self, node, options=None):
             if globalvar.is_constant:
                 # Global constants lose the constant value but are read-only
+                # TODO: When #633 and #11 are implemented, warn the user that
+                # it should transform the constants to literal values first.
```
s/it should/they should/

sergisiso replied:
Done

sergisiso commented:
@arporter I addressed inline comments and added the new transformation in the User Guide. This is ready for the next review.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -3785,3 +3785,105 @@ def validate(self, loop, options=None):
+        kernel call.
+
+        :param node: a kernel call.
+        :type nodes: :py:class:`psyclone.psyGen.CodedKern`
```
s/nodes/node/ (I only spotted this because the type information is missing in the auto-generated documentation in the user guide.)

sergisiso replied:
Fixed

arporter commented on doc/user_guide/transformations.rst:
```diff
@@ -244,6 +244,20 @@ can be found in the API-specific sections).
+
+####
+
+.. autoclass:: psyclone.transformations.KernelGlobalsToArguments
```
Please could you move this section so that it is alphabetical order with respect to the other transformations.

sergisiso replied:
Done

sergisiso commented:
@arporter I didn't notice the transformations were ordered alphabetically, now it should be fine. Ready for next review.

arporter commented:
All requested changes have been made.
All tests pass (including with compilation).
Updated documentation builds OK and looks fine.
